File: demo/src/main/scala/demo5/ErrorHandling2.scala
package demo5

import akka.actor.ActorSystem
import akka.stream.ActorFlowMaterializer
import akka.stream._
import akka.stream.scaladsl._
import scala.concurrent.duration._
import akka.stream.scaladsl.FlowGraph.Implicits._
import scala.concurrent.Await

object ErrorHandling2 extends App {

  implicit val system = ActorSystem("demo5")
  import system.dispatcher

  val decider: Supervision.Decider = {
    case _: ArithmeticException => Supervision.Resume
    case _                      => Supervision.Stop
  }

  // materializer with a custom supervision strategy
  implicit val mat = ActorFlowMaterializer(
    ActorFlowMaterializerSettings(system).withSupervisionStrategy(decider))

  val source = Source(0 to 5).map(100 / _)
  val result = source.runWith(Sink.fold(0)(_ + _))

  val res = Await.result(result, Duration.Inf)
  println(res) // this will print 228

}

File: java-demo/src/main/java/demo5/ErrorHandling2.java
package demo5;

import java.util.stream.IntStream;

import akka.actor.ActorSystem;
import akka.japi.function.Function;
import akka.stream.ActorMaterializer;
import akka.stream.ActorMaterializerSettings;
import akka.stream.Supervision;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;
import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;

public class ErrorHandling2 {
  public static void main(String[] args) throws Exception {
    ActorSystem system = ActorSystem.create("demo5");

    Function<Throwable, Supervision.Directive> decider = e -> {
      if (e instanceof ArithmeticException)
        return Supervision.resume();
      else
        return Supervision.stop();
    };

    ActorMaterializerSettings settings = ActorMaterializerSettings.create(system).withSupervisionStrategy(
        decider);
    // materializer with a custom supervision strategy
    ActorMaterializer materializer = ActorMaterializer.create(settings, system);

    IntStream stream = IntStream.range(0, 6);
    Source<Integer, ?> source = Source.from(() -> stream.iterator()).map(x -> 100 / x);
    Future<Integer> result = source.runWith(Sink.fold(0, (x, y) -> x + y), materializer);
    Integer res = Await.result(result, Duration.Inf());

    System.out.println(res); // this will print 228
  }
}